collector/lib/ProcfsScraper.cpp

```diff
@@ -349,27 +359,66 @@ bool ReadConnectionsFromFile(Address::Family family, L4Proto l4proto, std::FILE*
+                         UnorderedMap<ino_t, EndpointInfo>* listen_endpoints) {
+  UnorderedSet<Endpoint> all_listen_endpoints;
+
+  for (auto data : conn_lines) {
+    if (data.remote.port() == 0 && data.remote.address() == Address(0, 0, 0, 0)) {
+      all_listen_endpoints.insert(data.local);
+      if (data.inode && listen_endpoints) {
+        auto& endpoint_info = (*listen_endpoints)[data.inode];
+        endpoint_info.endpoint = data.local;
+        endpoint_info.l4proto = L4Proto::UDP;
+      }
     }
   }
 
-  {
-    FDHandle net_tcp6_fd = openat(dirfd, "net/tcp6", O_RDONLY);
-    if (net_tcp6_fd.valid()) {
-      FileHandle net_tcp6(std::move(net_tcp6_fd), "r");
-      success = ReadConnectionsFromFile(Address::Family::IPV6, L4Proto::TCP, net_tcp6, connections, listen_endpoints) && success;
-    } else {
-      success = false;
+  return true;
+}
+
+// Reads connection information from a file and saves the information to connections and listen_endpoints maps
+bool ReadConnectionsAndEndpointsFromFile(Address::Family family, std::FILE* f, L4Proto l4proto,
+                                         UnorderedMap<ino_t, ConnInfo>* connections, UnorderedMap<ino_t, EndpointInfo>* listen_endpoints) {
+  std::vector<ConnLineData> conn_lines;
+
+  if (!ReadConnLinesFromFile(f, family, &conn_lines)) return false;
+
+  if (l4proto == L4Proto::TCP) {
+    ProcessTcpConnectionsAndEndpoints(conn_lines, connections, listen_endpoints);
+  } else {
+    ProcessUdpEndpoints(conn_lines, listen_endpoints);
+  }
+
+  return true;
+}
+
+// GetConnectionsAndEndpoints reads all active connections (inode -> connection info mapping) for a given network NS, addressed by
+// the dir FD for a proc entry of a process in that network namespace.
+bool GetConnectionsAndEndpoints(int dirfd, UnorderedMap<ino_t, ConnInfo>* connections, UnorderedMap<ino_t, EndpointInfo>* listen_endpoints) {
+  bool success = true;
+
+  std::vector<std::tuple<std::string, L4Proto>> net_files_with_l4protocols{
+	  std::make_tuple("net/tcp", L4Proto::TCP),
+	  std::make_tuple("net/udp", L4Proto::UDP),
+	  std::make_tuple("net/udplite", L4Proto::UDP)
+  };
+
+  std::vector<Address::Family> address_families{Address::Family::IPV4, Address::Family::IPV6};
+
+  for (auto net_file_with_l4protocol : net_files_with_l4protocols) {
+    std::string net_file = std::get<0>(net_file_with_l4protocol);
+    L4Proto l4proto = std::get<1>(net_file_with_l4protocol);
+    for (Address::Family address_family : address_families) {
+      std::string net_file_ipv = (address_family == Address::Family::IPV4) ? net_file : net_file + "6"; 
+      FDHandle net_file_ipv_fd = openat(dirfd, net_file_ipv.c_str(), O_RDONLY);
+      if (net_file_ipv_fd.valid()) {
+        FileHandle net(std::move(net_file_ipv_fd), "r");
+        success = ReadConnectionsAndEndpointsFromFile(address_family, net, l4proto, connections, listen_endpoints) && success;
+      } else {
+        success = false;
+      }
```

Comment:
@Stringy suggested using list of tuples with (filename, family, protocol). I decided to go with a list of tuples of just (filename, protocol) and loop over family. The protocol can be obtained from the filename so it is not necessary to have the tuple and we could just have a vector of filenames. I still decided to go with the tuple to avoid calling a function that determines the protocol from the filename. I am happy make that change or any other improvement.

Note also that this will return false when any of the files are not there, which will be almost always.

Reply:
I don't think it is harmful that this function almost always returns false, but it is a bit odd. It looks if it returns false it checks if it can get the inode again, and if it can it does nothing. If it can't it erases some data.
```
     if (!GetConnectionsAndEndpoints(dirfd, &ns_network_data.connections, listen_endpoints ? &ns_network_data.listen_endpoints : nullptr)) {
          // If there was an error reading connections, that could be due to a number of reasons.
          // We need to differentiate persistent errors (e.g., expected net/tcp6 file not found)
          // from spurious/race condition errors caused by the process disappearing while reading
          // the directory. To determine if the latter is the root cause, we reattempt to read the
          // network namespace inode; if that succeeds, we assume that the process is still alive
          // and any errors encountered are persistent.
          uint64_t netns_inode2;
          if (!GetNetworkNamespace(dirfd, &netns_inode2) || netns_inode2 != netns_inode) {
            conns_by_ns.erase(emplace_res.first);
            continue;
          }
        }
```